dev/sg/internal/cloud/client.go

```diff
@@ -3,52 +3,129 @@ package cloud
 import (
 	"context"
 	"net/http"
+	"os"
+	"strings"
 
 	"connectrpc.com/connect"
 	cloudapiv1 "github.com/sourcegraph/cloud-api/go/cloudapi/v1"
 	"github.com/sourcegraph/cloud-api/go/cloudapi/v1/cloudapiv1connect"
 	"github.com/sourcegraph/run"
 
 	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/sourcegraph/lib/pointers"
 )
 
 const HeaderUserToken = "X-GCP-User-Token"
 const APIEndpoint = "https://cloud-ops-dev.sgdev.org/api"
+const EphemeralInstanceType = "internal"
+const DevEnvironment = "dev"
 
 type Client struct {
 	client cloudapiv1connect.InstanceServiceClient
 	token  string
+	email  string
 }
 
-func NewClient(ctx context.Context, endpoint string) (*Client, error) {
+type DeploymentSpec struct {
+	Name             string
+	Version          string
+	InstanceFeatures map[string]string
+}
+
+func NewDeploymentSpec(name, version string) *DeploymentSpec {
+	return &DeploymentSpec{
+		Name:    name,
+		Version: version,
+		InstanceFeatures: map[string]string{
+			"ephemeral": "true", // need to have this to make the instance ephemeral
+		},
+	}
+}
+
+func GetGcloudAccount(ctx context.Context) (string, error) {
+	return run.Cmd(ctx, "gcloud config get account").Run().String()
+}
+
+func NewClient(ctx context.Context, email, endpoint string) (*Client, error) {
+	err := validateEmail(email)
+	if err != nil {
+		return nil, err
+	}
+
+	// have to use IDENTITY token not ACCESS token!
+	token, err := run.Cmd(ctx, "gcloud auth print-identity-token").Run().String()
+	if err != nil {
+		return nil, errors.Newf("failed to get gcloud auth token: %v", err)
+	}
+
 	client := cloudapiv1connect.NewInstanceServiceClient(
 		http.DefaultClient,
 		endpoint,
 	)
 
-	token, err := run.Cmd(ctx, "gcloud auth print-identity-token").Run().String()
-	if err != nil {
-		return nil, errors.Newf("failed to get gcloud auth token: %v", err)
+	return &Client{client: client, email: email, token: token}, nil
+}
+
+func validateEmail(email string) error {
+	if len(email) == 0 {
+		return errors.New("gcloud account email is empty")
+	}
+
+	if !strings.Contains(email, "@sourcegraph.com") {
+		return errors.Newf("gcloud account email %q is not a valid Sourcegraph email", email)
 	}
 
-	return &Client{client: client, token: token}, nil
+	return nil
 }
 
 func newRequestWithToken[T any](token string, message *T) *connect.Request[T] {
 	req := connect.NewRequest(message)
 	req.Header().Add(HeaderUserToken, token)
+	req.Header().Add("Content-Type", "application/json")
 	return req
 }
 
-func (c *Client) ListInstances(ctx context.Context) ([]*cloudapiv1.Instance, error) {
-	req := newRequestWithToken(c.token, &cloudapiv1.ListInstancesRequest{})
-	res, err := c.client.ListInstances(
+func (c *Client) ListInstances(ctx context.Context) ([]*Instance, error) {
+	req := newRequestWithToken(c.token, &cloudapiv1.ListInstancesRequest{
+		InstanceFilter: &cloudapiv1.InstanceFilter{
+			AdminEmail: pointers.Ptr(c.email),
```

Comment:
nit: the purpose of `pointers.Ptr` is to be able to write `pointers.Ptr("666")`, here it's just adding noise. Not that this is really a problem, but I can see this propagating as an habit like the flu. 

```suggestion
			AdminEmail: &c.email,
```